File: cpp/20_valid_parentheses.cpp
#include <stack>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  bool isValid(string s) {
    unordered_map<char, char> paren = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
    stack<char> stack;
    for (const auto &c : s) {
      if (paren.count(c) > 0) {
        stack.push(c);
      } else {
        if (stack.empty() || paren[stack.top()] != c) {
          return false;
        }
        stack.pop();
      }
    }
    return stack.empty();
  }
};

File: rust/20_valid_parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
        for c in s.chars() {
            match c {
                '(' | '{' | '[' => stack.push(c),
                _ => match stack.pop() {
                    Some('(') if c == ')' => (),
                    Some('{') if c == '}' => (),
                    Some('[') if c == ']' => (),
                    _ => return false,
                },
            }
        }

        stack.is_empty()
    }
}
